### DISCUSSION TITLE
Manually check whether an image is out of date

### DISCUSSION THREAD

**Discussion by SimeonEhrig** [containers/podman#24649] [Q&A]
Hello,

I have a special case to update a container. If I update a Nextcloud container, a new image is downloaded. When the new image is starting, it will check the container version and the version of the Nextcloud installation, which is stored in a persistent volume. If the installation is older than the container version it will automatically update the installation. An updated installation cannot be rolled back. 

Therefore, if the container fails to start after the container update, but already started the update process of the installation (maybe this process has a problem), Podman will automatically roll back the container, the old container will start and fail because a container with an older Nextcloud version cannot run with an installation of a newer version. So the roll back is also broken.

I solve the problem with following approach. During the backup process I do the following stuff:
1. create a backup
2. check for a new image by a hand-written, a little bit in-stable script
3. if the there is a new image, tag the current image and the backup with special tags, that I know how to restore the backup
4. upgrade the image manually

The problem is the handwritten script for checking new images. Podman has already the mechanism. Therefore it would be cool if I could use the mechanism e.g. `podman check-update <image_name>`.

Does somebody know if this is already possible? 

A workaround would be to create an quadlet with a Nextcloud image and the auto update property and then run `podman auto-update --dry-run` but the start up time for a new Nextcloud container takes a while, because it does download a whole Nextcloud installation.

**Comment by debarshiray** [comment]
> The problem is the handwritten script for checking new images. Podman has already the mechanism. Therefore it would be cool if I could use the mechanism e.g. `podman check-update <image_name>`.

Did you already try:
```ShellSession
$ podman inspect --format '{{ .ImageDigest }}' --type container CONTAINER
sha256:abcd
$ skopeo inspect --format '{{ .Digest }}' docker://registry.url.to/image
sha256:1234
```

If the outputs are different, then you can assume that there's a newer version of the image on the registry.

**Reply by SimeonEhrig** [reply]
Thank you for your answer. In the meantime I develop the following solution.

```bash
#!/usr/bin/env bash

REPOSITORY=library/nextcloud
TAG=31

TOKEN=$(curl -s "https://auth.docker.io/token?service=registry.docker.io&scope=repository:$REPOSITORY:pull" | jq -r ".token")
REMOTE_DIGEST=$(curl -s -H "Authorization: Bearer $TOKEN" -H "Accept: application/vnd.docker.distribution.manifest.v2+json" https://index.docker.io/v2/$REPOSITORY/manifests/$TAG |
    jq -r '.manifests.[] | select(.platform.os=="linux" and .platform.architecture=="amd64") | .digest')

LOCAL_REMOTE_DIGESTS=$(podman image inspect --format json $REPOSITORY:$TAG | jq '.[].RepoDigests.[]')

for remote_digest in $LOCAL_REMOTE_DIGESTS; do
    if echo $remote_digest | grep -s "$REMOTE_DIGEST" 2>&1 >/dev/null; then
        echo "$REPOSITORY:$TAG up to date"
        exit 0
    fi
done

echo "$REPOSITORY:$TAG is out of date"
```

I directly pull the image sha from the docker hub rest API. If the sha is in the `RepoDigests` sha list the image is up to date. Otherwise I need to pull a new image. This works stable. I tried your solution, but I cannot find the sha from skopeo in the container. But the sha is also in the `RepoDigests` list. Looks like I could replace my API request with `scorpio`.

```bash
#!/usr/bin/env bash

REPOSITORY=library/nextcloud
TAG=31

REMOTE_DIGEST=$(skopeo inspect --format '{{ .Digest }}' docker://docker.io/$REPOSITORY:$TAG)

LOCAL_REMOTE_DIGESTS=$(podman image inspect --format json $REPOSITORY:$TAG | jq '.[].RepoDigests.[]')

for remote_digest in $LOCAL_REMOTE_DIGESTS; do
    if echo $remote_digest | grep -s "$REMOTE_DIGEST" 2>&1 >/dev/null; then
        echo "$REPOSITORY:$TAG up to date"
        exit 0
    fi
done

echo "$REPOSITORY:$TAG is out of date"
```

This already helps a lot, because I my request was motivated because docker hub changed the API. It takes me more than a day to fix it and podman already know how to use the new API. Unfortunately I didn't find the place in the code where the check happen. Therefore I thought the easy solution is ask for the feature. But skorpio also solves the API problem for me :-) Thanks
